src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -231,39 +235,183 @@ def frobenius_charpoly(self, var='X'):
+        works for Drinfeld modules defined over Fq[T], but otherwise
+        does not impose any other constraints, including on the rank,
+        minimal polynomial, or that the Drinfeld module is defined
+        over the prime field.
+
+        INPUT:
+
+        - ``var`` (default: ``'X'``) -- the name of the second variable
 
-            We compute the Frobenius norm, and with it the Frobenius
-            trace. This gives the Frobenius characteristic polynomial.
-            See [MS2019]_, Section 4.
+        EXAMPLES::
 
-            See docstrings of methods :meth:`frobenius_norm` and
-            :meth:`frobenius_trace` for further details on the
-            computation of the norm and of the trace.
+            sage: Fq = GF(25)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(8)
+            sage: phi = DrinfeldModule(A, [z, 4, 1, z, z+1, 2, z+2, 1, 1, 3, 1])
+            sage: phi._frobenius_charpoly_crystalline()
+            X^10 + X^9 + (3*T + z2 + 1)*X^8 + (4*T^2 + z2*T + 2*z2 + 1)*X^7 + ... + (4*z2 + 4)*T^4 + 4*z2*T^2 + (z2 + 2)*T + z2
+
+            sage: Fq = GF(27)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(10)
+            sage: phi = DrinfeldModule(A, [z, z^2 + z, 2, 1, z, z+1, 2, z+2, 0, 1, 1, z^2 + z])
+            sage: phi._frobenius_charpoly_crystalline()
+            X^11 + (z3^2 + 2*z3)*X^10 + ((z3 + 1)*T + z3)*X^9 + ((2*z3^2 + z3 + 2)*T^2 + ... + (2*z3^2 + 2*z3 + 2)*T + z3^2
+
+        ALGORITHM:
+
+        Compute the characteristic polynomial of the Frobenius endomorphism
+        acting on the crystalline cohomology of a Drinfeld module, which
+        is equal to that of the Frobenius endomorphism on the Drinfeld
+        module. A recurrence on elements of the cohomology allows us to
+        compute a matrix representation of the Frobenius endomorphism
+        efficiently using a companion matrix method.
         """
-        self._check_rank_two()
-        A = self._function_ring  # Fq[T]
-        S = PolynomialRing(A, name=var)  # Fq[T][X]
-        # Does not work when Fq is not a prime field...
-        # chi = self._gen.reduced_charpoly()
-        # return -chi(A.gen(), S.gen())
-        return S([self.frobenius_norm(), -self.frobenius_trace(), 1])
+        Fq, L = self._Fq, self.base_over_constants_field()
+        A = self.function_ring()
+        q, r, n = Fq.cardinality(), self.rank(), L.degree(Fq)
+        nstar = ceil(sqrt(n))
+        n1, n0 = n // nstar, n % nstar
+        dm = self.coefficients(sparse=False)
+        S = PolynomialRing(L, name=str(A.gen()))
+        SM = MatrixSpace(S, r, r)
+        mu_coeffs = ((S.gen() - dm[0])**(n+1)).coefficients(sparse=False)
+
+        def companion(order):
+            # + [1] is required to satisfy formatting for companion matrix
+            M = SM(companion_matrix([(dm[i]/dm[r])**(q**order)\
+                    for i in range(r)] + [1], format = 'top'))
```

Comment:
1. Removed the useless `\`.
2. Reflowed.
3. No spaces before and after `=` when dealing with default arguments in a function.

```suggestion
            M = SM(companion_matrix([(dm[i]/dm[r])**(q**order)
                                    for i in range(r)] + [1], format='top'))
```